src/goto-programs/remove_instanceof.cpp

```diff
@@ -42,160 +42,129 @@ class remove_instanceoft
 
   bool lower_instanceof(goto_programt &);
 
-  typedef std::vector<
-    std::pair<goto_programt::targett, goto_programt::targett>> instanceof_instt;
+  bool lower_instanceof(goto_programt &, goto_programt::targett);
 
-  void lower_instanceof(
-    goto_programt &,
-    goto_programt::targett,
-    instanceof_instt &);
-
-  void lower_instanceof(
-    exprt &,
-    goto_programt &,
-    goto_programt::targett,
-    instanceof_instt &);
-
-  bool contains_instanceof(const exprt &);
+  int lower_instanceof(exprt &, goto_programt &, goto_programt::targett);
 };
 
-/// Avoid breaking sharing by checking for instanceof before calling
-/// lower_instanceof.
-/// \par parameters: Expression `expr`
-/// \return Returns true if `expr` contains any instanceof ops
-bool remove_instanceoft::contains_instanceof(
-  const exprt &expr)
-{
-  if(expr.id()==ID_java_instanceof)
-    return true;
-  forall_operands(it, expr)
-    if(contains_instanceof(*it))
-      return true;
-  return false;
-}
-
-/// Replaces an expression like e instanceof A with e.@class_identifier == "A"
-/// Or a big-or of similar expressions if we know of subtypes that also satisfy
+/// Replaces an expression like e instanceof A with e.\@class_identifier == "A"
+/// or a big-or of similar expressions if we know of subtypes that also satisfy
 /// the given test.
-/// \par parameters: Expression to lower `expr` and the `goto_program` and
-/// instruction `this_inst` it belongs to.
-/// \return Side-effect on `expr` replacing it with an explicit clsid test
-void remove_instanceoft::lower_instanceof(
+/// \param expr: Expression to lower (the code or the guard of an instruction)
+/// \param goto_program: program the expression belongs to
+/// \param this_inst: instruction the expression is found at
+/// \return number of instanceof expressions that have been replaced
+int remove_instanceoft::lower_instanceof(
```

Comment:
Can this be negative?

Reply:
No. Personally I favour the signed type though because if you subtract one from another and the answer would be negative you don't unexpectedly underflow and get a very large answer.

Reply:
Are you otherwise happy to merge? Or did you comment rather than approve because you have other concerns?

Reply:
Likely a std::size_t